src/vs/workbench/api/common/extHostLanguageFeatures.ts

```diff
@@ -1084,41 +1099,40 @@ class InlineCompletionAdapter extends InlineCompletionAdapterBase {
 				if (item.command) {
 					if (!disposableStore) {
 						disposableStore = new DisposableStore();
-						this._disposables.set(pid, disposableStore);
 					}
 					command = this._commands.toInternal(item.command, disposableStore);
 				}
 
-				const insertText = item.insertText ?? item.text;
-				if (insertText === undefined) {
-					throw new Error('text or insertText must be defined');
-				}
+				const insertText = item.insertText;
 				return ({
 					insertText: typeof insertText === 'string' ? insertText : { snippet: insertText.value },
+					filterText: item.filterText,
 					range: item.range ? typeConvert.Range.from(item.range) : undefined,
 					command,
 					idx: idx,
-					completeBracketPairs: item.completeBracketPairs
+					completeBracketPairs: this.isAdditionProposedApiEnabled ? item.completeBracketPairs : false
 				});
 			}),
+			commands: commands.map(c => {
+				if (!disposableStore) {
+					disposableStore = new DisposableStore();
+				}
+				return this._commands.toInternal(c, disposableStore);
+			})
 		};
 	}
 
 	override disposeCompletions(pid: number) {
-		this._cache.delete(pid);
-		const d = this._disposables.get(pid);
-		if (d) {
-			d.clear();
-		}
-		this._disposables.delete(pid);
+		const data = this._references.disposeReferenceId(pid);
+		data?.dispose();
 	}
 
 	override handleDidShowCompletionItem(pid: number, idx: number): void {
-		const completionItem = this._cache.get(pid, idx);
+		const completionItem = this._references.get(pid)?.items[idx];
 		if (completionItem) {
-			InlineCompletionController.get(this._provider).fireOnDidShowCompletionItem({
-				completionItem
-			});
+			if (this._provider.handleDidShowCompletionItem && this.isAdditionProposedApiEnabled) {
+				this._provider.handleDidShowCompletionItem(completionItem);
+			}
 		}
```

Comment:
When registering an inline completion item provider you can pass along to the renderer if the provider supports `handleDidShow...`. With that you can skip the IPC call for most providers. See `this._proxy.$registerCompletionsProvider` and others for sample cases

Reply:
Should I just check for `this.isAdditionProposedApiEnabled` or if `handleDidShowCompletionItem` is set?
The field `handleDidShowCompletionItem` could be initialized later.

Reply:
> The field handleDidShowCompletionItem could be initialized later.

We ignore that and for consistence reasons should continue to do so. If it's not there at `register`-time than it is never there